packages/material-ui/src/CssBaseline/CssBaseline.test.js

```diff
@@ -1,30 +1,27 @@
 import React from 'react';
 import { assert } from 'chai';
-import { createMount, findOutermostIntrinsic } from '@material-ui/core/test-utils';
+import { createMount } from '@material-ui/core/test-utils';
 import CssBaseline from './CssBaseline';
 
 describe('<CssBaseline />', () => {
   let mount;
 
   before(() => {
-    mount = createMount();
+    // StrictModeViolation: makeStyles will retain the styles in the head in strict mode
+    // which becomes an issue for global styles
+    mount = createMount({ strict: false });
```

Comment:
/cc @oliviertassinari 

https://codesandbox.io/s/k2lpvz5745: 
- works only on the first run which seems to break everything
- After unmounting styles are removed as expected within a fragment
- if mounted in strict mode styles are retained after unmounting. This can be observed by enabling the strict mode test for CssBaseline and running `test:karma`. Likely due to double render invocations.

Reply:
Thanks for the reproduction. I believe it simulates what happens when the render phase is interrupted and resumed. We lose track of what was injected in the page, the stylesheet is not removed. On a side note, I believe styled-components has the same problem: https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/constructors/createGlobalStyle.js.

It sounds like we are not done yet with the styling solution…

https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects

Reply:
Essentially. Strict mode just runs the render twice before committing which will happen in concurrent mode. We attach in the render phase but detach in the cleanup phase (which is tied to a commit). We can end up attaching multiple times but only detaching once.

We should clean makeStyles up so that we can put as much as possible in actual effects. We should end up with only registering the sheet in a render phase side effect. For client side react we can downgrade this to a layout effect which makes it safe. Everything else should be put in their proper effects.

Reply:
It renders twice before committing. Since we attach in render but detach in a commit cleanup we leak.

We need to cleanup the implementation to put every side effect into their proper hook. Only registering the sheet should be put into an effect that we upgrade on the server to a render effect.

Reply:
Yes, we need to separate the creation of the stylesheet and its injection in the dom. The creation of the style sheet should happen in the render phase. The attack of the style sheet should happen in a layout effect.

Reply:
I mean we don't outright have to attack it. Attaching might scare the CSS enough :smile:

Do we need to create the sheet on client during render as well?

Reply:
We need the class names in the render phase.